core/commonMain/src/kotlinx/io/Output.kt

```diff
@@ -113,32 +88,47 @@ abstract class Output(bufferSize: Int = DEFAULT_BUFFER_SIZE) : Closeable {
     }
 
     /**
-     * Write [source] buffer to destination.
+     * Request to fill [buffer] from [startIndex] to [endIndex] exclusive To destination.
      *
-     * May block until destination has no available space.
+     * This method won't modify [source] and will block until all bytes from [source] won't be flushed.
      */
-    protected abstract fun flush(source: Buffer, length: Int)
+    protected abstract fun flush(source: Buffer, startIndex: Int, endIndex: Int)
+
+    /**
+     * Closes the underlying source of data used by this output.
+     * This method is invoked once the output is [closed][close].
+     */
+    protected abstract fun closeSource()
 
     private fun flushBuffer() {
-        flush(buffer, position)
-        buffer = bufferPool.borrow()
+        flush(buffer, 0, position)
+        buffer = pool.borrow()
         position = 0
     }
 
-    internal inline fun writeBufferRange(writer: (buffer: Buffer, startOffset: Int, endOffset: Int) -> Int) {
-        var startOffset = position
-        var endOffset = buffer.size - 1
-        if (startOffset > endOffset) {
+    /**
+     * Call [writer] block with [buffer] and write [startOffset]
+     *
+     * The [writer] expected to return new [buffer] position.
+     *
+     * @return number of written bytes.
+     */
+    internal inline fun writeBuffer(writer: (buffer: Buffer, bufferStart: Int, bufferEnd: Int) -> Int): Int {
+        if (position == buffer.size) {
             flushBuffer()
-            startOffset = position
-            endOffset = buffer.size - 1
         }
 
-        val newPosition = writer(buffer, startOffset, endOffset)
+        val newPosition = writer(buffer, position, buffer.size)
+        val result = newPosition - position
         position = newPosition
+        return result
+    }
+
+    internal fun size(): Int {
+        return position
     }
 
-    private inline fun writePrimitive(
+    internal fun writePrimitive(
         primitiveSize: Int,
         writeDirect: (buffer: Buffer, offset: Int) -> Unit,
         longValue: () -> Long
```

Comment:
We should get rid of this lambda on the hot path of writing a primitive

Reply:
I'll revert `inline` remove here and keep lambda. I want to measure the experiment with early `flush` and `buffer.write` instead of `byte-by-byte` copy.